File: code-monkey/src/tools/write_file_tool.py
import os
from pydantic import BaseModel, Field
from typing import Type, Optional
from langchain_core.callbacks import (
    AsyncCallbackManagerForToolRun,
)
from .io_tool import IOTool
from .utils import make_file_path
from instrumentation import instrument

class WriteFileToolInput(BaseModel):
    fname: str = Field(description="Name of the file to edit.")
    content: str = Field(description="New contents of the file.")

class WriteFileTool(IOTool):
    """Tool to overwrite a file of given name with passed-in content"""
    name: str = "write_file"
    description: str = "Write content to the file of given name"
    args_schema: Type[BaseModel] = WriteFileToolInput

    @instrument("Tool._run", ["fname", "content"], attributes={ "tool": "WriteFileTool" })
    def _run(self, fname: str, content: str, run_manager: Optional[AsyncCallbackManagerForToolRun] = None) -> None:
        file_path = make_file_path(fname)
        os.makedirs(os.path.dirname(file_path), exist_ok=True)
        with open(file_path, "w") as file:
            file.write(content)
        self.notify_file_modified(fname)
        return None

File: code-monkey-ts2/tools/write_file_tool.ts
import * as fs from 'fs';
import * as path from 'path';
import { z } from "zod";
import { makeFilePath } from './utils';
// TODO import { instrument } from '../instrumentation';
import { IOTool } from './io_tool';

const schema = z.object({
    fname: z.string().describe("Name of the file to edit."),
    content: z.string().describe("New contents of the file."),
});

export class WriteFileTool extends IOTool {
    name = "write_file";
    description = "Write content to the file of given name";
    schema = schema;


    // TODO @instrument("Tool._call", ["fname", "content"], { tool: "WriteFileTool" })
    async _call({ fname, content }: z.infer<typeof schema>): Promise<string> {
        try {
            const filePath = makeFilePath(fname);
            fs.mkdirSync(path.dirname(filePath), { recursive: true });
            fs.writeFileSync(filePath, content);
            await this.notifyFileModified(fname);
            return "file successfully written";
        } catch (error) {
            console.error(`Failed to write file: ${fname}`);
            console.error(error);
            return "failed to write file";
        }
    }
}